src/transformStateWithClones.js

```diff
@@ -6,8 +6,38 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentState = Object.assign({}, state);
+
+  for (const value of actions) {
+    switch (value.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = Object.assign({}, currentState, value.extraData);
+
+        break;
+
+      case 'removeProperties':
+        for (const key of value.keysToRemove) {
+          if (currentState.hasOwnProperty(key)) {
+            delete currentState[key];
+          }
+        }
+
+        break;
```

Comment:
it's better not to add extra empty lines to switch/case to avoid confusion about which belongs to which
```suggestion
      case 'addProperties':
        currentState = Object.assign({}, currentState, value.extraData);
        break;

      case 'removeProperties':
        for (const key of value.keysToRemove) {
          if (currentState.hasOwnProperty(key)) {
            delete currentState[key];
          }
        }
        break;
```